### ISSUE TITLE
New view for series

### ISSUE THREAD

**Issue by gpapin** [janeczku/calibre-web#979]
Hi, I've been working on a quick proof of concept to improve the listing of the series.
Basically instead of having a list of series titles I would have something like this:

<img width="1438" alt="Screenshot 2019-07-27 12 23 14" src="https://user-images.githubusercontent.com/3064311/61993808-655a9280-b069-11e9-93a4-1d51de343713.png">

Now in order to integrate it smoothly I want to have your opinion as to how this should behave.
Here are my thoughts:

**1: Having another menu called "Collection" next to serie witch would display the new view**
> ➖A bit redundant 
> ➕You get to keep both view and it's easy to switch between them

**2: Having a switch on top of the serie page to switch between views**
> ➕Easy to switch between views
> ➖If you prefer one way or another you'll be stuck with what the default would be
> ➖Extra ui

**3: Having a setting on your user pref to choose between one view or another.**
> ➕Once setup, no extra step or cluttering to get your view.
> ➖More difficult to switch between different style

My personal preference goes for option 3, but I would like to hear your thoughts on that.

Also Things to consider and features I'd like to add to this view.

**cover selection**:
At the moment I just display the cover of the first book in the collection. We could imagine an option to select  
* First
* Last
* First unread

**Badge display**:
At the moment I don't display any badge but there is going to be one and it could display 2 things: 
* Total books in the collection
* Total unread books in the collection

**Comment by gpapin** [comment]
I just discovered the plex-like theme and using it breaks thinks a little, I would like to fix it but as far as I can see it's in the caliBlur.min.css and there is no sass or non-minified version ?
How do you work with your css @OzzieIsaacs ?

**Comment by OzzieIsaacs** [comment]
Regarding question one, the different viewing options. I would go for option 2, we could store the selection in the backend. At opening the page again we could load the stored setting (similar to the read/unread flag)
Cover selection: I know somebody will ask for different options, but like Jürgen Klopp said: "the life is not a wish concert" (Life is no a picnic), so we should make it not to complicate and leave it with the cover of the first book (many people open issues here because they don't find certain options in the settings).
Badge: Difficult question, there is an issue ( #422) which is requesting previews for series, authors and tags. I think this request could be solved the way you are proposing. For an author view I would more expect total books. I tend a little bit to total books badge.

**Comment by OzzieIsaacs** [comment]
@leram84: Do you have published the non minified version of the caliBlur.min.css somewhere?

**Comment by OzzieIsaacs** [comment]
Okay, I think, found it: https://github.com/leram84/layer.Cake/blob/master/caliBlur/caliBlur.css

**Comment by gpapin** [comment]
I can always go back to option 2 but I've already finished coding option 3 - ie stored in the user settings

<img width="1431" alt="Screenshot 2019-07-27 17 28 12" src="https://user-images.githubusercontent.com/3064311/61997080-0d855100-b094-11e9-9023-a79ebbf96194.png">

If you want to go back to display this switch in the view itself it's fine but I can't find the read/unread option you are talking about.

For the cover selection I agree with you, this is going to be too complicated and we have to come up with something sensible. My preferred approach was to display the latest non read book, but it turns out it's a little too complicated to implement so I ended up displaying the latest of the serie. In my opinion, a bit better than the first because if you start a serie from the beginning, you can see the thumbnail evolve over time.

Same thing with the count badge, I ended up just showing the total number of books in the serie.
Basically having the read value in a custom table makes it difficult to work with as it means you have to join another table - As it's the first time I use sqlalchemy it's a bit over my knowledge right now.

Regarding having a cover view for author and tag, I' not so convinced right now. In my opinion here a list of word works better. Despite having a cover would mean to be able to upload and store custom image for it - something that isn't supported right now. On top of that, from a user perspective you would have to find a picture of every single author which could be troublesome, but worst, how do you find images that illustrate a tag, unless you make them yourself, it's going to be very hard to find something that will match the theme...

**Comment by gpapin** [comment]
Thinking on the display mode, maybe we could do something similar to what lex does (picture attached for reference) that said not sure where and how that would fit in the base theme...

<img width="443" alt="Screenshot 2019-07-27 17 56 48" src="https://user-images.githubusercontent.com/3064311/61997341-2bed4b80-b098-11e9-9dd1-1f5c0e7c328c.png">

**Comment by OzzieIsaacs** [comment]
Read/unread: Maybe I thought to complicated
![image](https://user-images.githubusercontent.com/6438895/61997374-12043680-b0a1-11e9-9e78-be58fe9dd990.png)
You just tick the checkbox and it's stored in the background via ajax call, so the user doesn't really recognize that he did change a setting. And this setting also has to be stored per user base.
Regarding author page, you are right, I didn't think really through it.

**Comment by gpapin** [comment]
Right I'm going to start working on porting this setting on the serie view itself. But I won't be able to do it via an ajax call because the app will have to refresh the view. On a grid mode I don't load the `list.html` and I have to make a different SQL query to get the book info associated with the Serie so really it needs a full refresh.

The way I see it right now is having this : 
<img width="233" alt="Screenshot 2019-07-27 18 25 17" src="https://user-images.githubusercontent.com/3064311/61997618-f34f7100-b09b-11e9-92c6-5e23fa035fee.png">  

When you click it, it sends an ajax post to the `/me` page (where all personal preferences are handled) and on success refresh the page.

**Comment by OzzieIsaacs** [comment]
Maybe it would be easier on clicking first or second option you trigger a page refresh (call url-A or url-B) and toggle the setting if necessary.

**Comment by gpapin** [comment]
@OzzieIsaacs & @leram84  So regarding Caliblur theme how do you see it working in the future ?

Is this something that should really leave outside of calibre-web repo ? For me, now that it's fully integrated it would make more sense to put all the css here so contributor can edit the theme as they add new feature. This means getting rid of the minified version for now, we could minify it on the go when serving the file but the minified version shouldn't be in the repo, it's like a binary version of the css.

Right now the easiest way for me to support the theme is copying the css file linked above, edit it and serve this file instead of the minified version.

Alternatively I could serve another file on top of the minified version to override some of the things and add more things, but not ideal as it gives less control...

**Comment by leram84** [comment]
Hey guys. I've just been swamped w irl stuff lately and haven't had time to dig into the latest changes. i have some time this week so should be able to bang out some of the new features, and take a look at this series stuff. Philosophically, im open to anyone using my work however they want and using an unminified version of the css would probably make that easier. 

As a rule i never leave anything over 5k lines unminified in production, but realistically its not going to add that much overhead to this app, so feel free.... but the real problem is a little more complicated. This all started as a completely seperate addon project and therefor is not done in the most... rational way lol. I honestly don't really think anyone else can edit it at this point without introducing a series of bugs all over the place lol. The css was done by hand, element by element, and is a bit of a house of cards at this point. 

If you would like to go forward, i would suggest you create a completely new file to add on instead of editing anything in mine. It will save your sanity and let you track down bugs much easier. Otherwise, assuming i can bang out the other issues i have in the queue, i can hopefully get to this soon. The only thing is, i would prefer to work on this when the entire feature is completely finished with the base theme. Its a lot of work to change things so its easier to just wait till its all done and then "blur" it all at once lol,

**Comment by gpapin** [comment]
That makes complete sense. To be fair the feature works and it doesn't look more broken that it currently is with Caliblur theme. We can then focus on making the filters integrate nicely.

On the CSS topic, I see where you are coming from, working with CSS always end up in a mess...
I would be keen to have some sass integration but that's maybe too much to ask. At the moment I have 20 lines of css in a `caliblur_override.css` file so no big deal.

I'm pretty busy this week so I won't be able to help much, but I'll try to submit something tonight so you guys can take a look and add feedback.

**Comment by leram84** [comment]
no worries. i have some open issues i need to bang out first so theres no rush on my end. And as for this actual feature btw, im a big fan of this! A few thoughts from my (caliblur) perspective:

1) My vote is #2 as well. having the switch from grid to list be available on each page (like plex) is much more intuitive imo, And this isn't a different list right? so calling one collections just because it has images and the other series because its text feels even more complicated. 

2) For the badge, it kind of depends. Right now, there is no badge for read/unread, so i agree that adding that concept to the badge in relation to number of books in a series won't be intuitive. I DO however think it would make sense to do `total unread books` instead of just `total books` ONLY if you are planning on adding a read/unread badge to each book in ever view of the interface (again, this is in line w plex). So in other words my vote is currently to do total books only, but if you ever add the read/unread badge to the main pages, then to change this to total unread books.

3) As for the author thumbs, I dont think it matters that much, but i would actually say its doable. Author images can be automated from goodreads, and you do actually get those images on the specific author page right now if you have them. I made caliblur work with both, and could do the same with thumbnails if you add it. If an image does't exist it will just show the default author icon.... I'm betting eventually someone who doesn't have any author images and doesn't feel like setting that up w goodreads might eventually ask for a server side switch to disable the grid view toggle from the page for their users, but 🤷‍♂ 

Anyway, this is all just my 2 cents.

**Comment by gpapin** [comment]
1. So yes it's the same list, but different SQL query as I need a bit more info to actually display a cover. The idea of option 3 was to have this same switch but permanently in the user user pref section because that was less work for me :) - but realistically it's better to have it in the view itself, and we keep the benefit of having the permanent settings as it will remember your choice each time you visit that page. Anyways, option 2 is implemented now so everyone's happy.

2. Yeah having the read/unread badge is something I would be keen on seing implemented but it's a bit of a pain in the ass. Right now it's stored in a custom_colum table in calibre *or* if you don't haven't set that up, then it stores it in the calibre-web own database making it impossible to join... Like you said it would be nice to have and I agree on what you said but maybe later as it's a bit out of scope as it opens up a whole new topic. For now it display the total number of books to stay consistent with what the current list show you.

3. I'm from those who don't have any author images, so not a huge priority for me right now. That said this cover feature could set the foundation and it could be expanded to other views later on.

**Comment by gpapin** [comment]
PR open if you want to take a look: https://github.com/janeczku/calibre-web/pull/986

**Comment by Knepherbird** [comment]
A little while back, when I discovered the goodreads option for pulling author info, I was thinking the Authors page would look very nice in a responsive grid layout with the goodreads photos (with a generic thumb as a fallback?). I tinkered at the time in firefox's dev inspector with some css grid for layout but never saved my tweaks. When I updated my book library today, I saw reference to some changes for comic book functionality, and decided to get the latest and point it at my comic book library, and there's definitely some promise in calibre-web being a better solution than ubooquity for me. But that Series page definitely needs some visual love, and I'm happy there's work being done on a grid view. 

One thing I'd wonder is, while an ajax call would useful for setting a user's preference to remember the next time the page is visited, couldn't the toggle fire an event to change a class on the outer container and let css handle switching between grid and list views in the browser?

**Comment by gpapin** [comment]
It's not as simple as adding a css class to change the layout, because SQL query to get the information for the list view is not the sae as the one fo the grid view. Basically in the list view you don't get the books just the series so we can't display the picture. The solution to this issue would be to get all the data regardless of the type of view and then hide the irrelevant data in the list view, but I didn't go that route because the list view is a generic template for all the page that support it not just the series.

**Comment by ameel** [comment]
hello,

is there any update on this view request? I would very much like to test/contribute

**Comment by eloscarfr** [comment]
Hello,
Is there a way to test this very promising theme add-on?
Thanks

**Comment by OzzieIsaacs** [comment]
You could try the development branch (it‘s merged there)

**Comment by eloscarfr** [comment]
Indeed, thanks a lot!

**Comment by eloscarfr** [comment]
Hi,
Sorry, I did installed the dev branch, but still cannot see how to display the Collection folder (with covers). Could-you please help me ?
Thanks

**Comment by Knepherbird** [comment]
Didn't realize this had been implemented in the Develop branch, I'd been playing around with some different ways of implementing something like this (primarily on the html/css side). @eloscarfr, you should be able to see a "Grid" button at the end of the sort/filter buttons at the top of the Series page.

One thing I'm noticing is that resizing the browser causes a lag in resizing/reflowing the grid. Is it using some sort of javascript library for that? Has any consideration been given to moving away from bootstrap for this view and using CSS grid or Flexbox to handle the grid layout?

**Comment by eloscarfr** [comment]
Unfortunately, I don't see such a button... I'm on version "0.6.8 Beta" (from Develop branch), is it the one that includes the collection folder ?
Thanks.

**Comment by Knepherbird** [comment]
I may be making an incorrect assumption, if so I apologize. By "collection" are you referring to the "series" view? If so, I installed from source using git clone --branch Develop ..., and it does show in admin as 0.6.8 Beta. I'm attaching a screenshot of what is shown on the default Series page.
![Screenshot_2020-05-19 Calibre-Web Series](https://user-images.githubusercontent.com/5818504/82375542-9976b980-99d5-11ea-9a19-928a43cd8a01.png)

**Comment by Knepherbird** [comment]
I may be making an incorrect assumption, if so I apologize. By "collection" are you referring to the "series" view? If so, I installed from source using git clone --branch Develop ..., and it does show in admin as 0.6.8 Beta. I'm attaching a screenshot of what is shown on the default Series page.
![Screenshot_2020-05-19 Calibre-Web Series](https://user-images.githubusercontent.com/5818504/82375542-9976b980-99d5-11ea-9a19-928a43cd8a01.png)


> Unfortunately, I don't see such a button... I'm on version "0.6.8 Beta" (from Develop branch), is it the one that includes the collection folder ?
> Thanks.

I just glanced back at the original proposal above and I now see the confusion. Instead of implementing a separate "Collection" view as originally proposed, this now has an option in the Series view that can toggle between list and grid views.

**Comment by eloscarfr** [comment]
Hi, 
I may have an issue with my eyes :) but I definitively cannot see the button... see screenshoot below.
<img width="1174" alt="Capture d’écran 2020-05-21 à 15 49 35" src="https://user-images.githubusercontent.com/64691150/82565605-fc558580-9b7a-11ea-9824-0e8473058470.png">
Also, I've checked the version installed, and it seems to be the last one (nighty build), see below...
<img width="940" alt="Capture d’écran 2020-05-21 à 15 49 43" src="https://user-images.githubusercontent.com/64691150/82565667-1727fa00-9b7b-11ea-9ced-da9a47a2491a.png">

Any idea?
Thanks a lot.

**Comment by OzzieIsaacs** [comment]
The view is meanwhile merged to the master branch, a few minutes ago, I found the problem with the invisible button (was only visible with English language, or translated Series text is also Series)

**Comment by OzzieIsaacs** [comment]
So no problems with your eyes

**Comment by eloscarfr** [comment]
Finally I got it! I now have installed the main branch, and it works like a charm.
Congrats and thank you very much for this very useful feature.

**Comment by OzzieIsaacs** [comment]
Merged to master

**Comment by brahma-dev** [comment]
@gpapin  It is mentioned earlier that it shows cover of the latest book of the series. But it appears to be showing the cover of the first book of the series. Does it need some extra configuration ?

**Comment by gpapin** [comment]
@brahma-dev I took a quick look and it's working  out of the published date rather than the book index. I don't remember if the index serie wasn't there before but I would have though  the 2 would match.

I'd recommend you double check your book metadata.

**Comment by brahma-dev** [comment]
@gpapin The metadata is correct, also the series index. I tried changing them but the cover remained the same. So, I created another dummy series and it looks like it selects the cover for the book that was uploaded first, regardless of the metadata. Like if I add three book published in February, January and March, it will show the cover for the February one.

**Comment by brahma-dev** [comment]
In continuation with the above, I removed February one from the Series. The cover changed to January. Then I added February back to the Series. The cover stayed on January.

**Comment by gpapin** [comment]
That sounds like a bug by the looks of it.
It's been a while but I probably can pick it up if no-one does before me.

**Comment by OzzieIsaacs** [comment]
The newest nightly version should show the last book cover of the series (maybe again)

**Comment by bcsbugsy** [comment]
Since Version 0.6.19 the oldest cover-version of a series appears in the series grid-view 👎

**Comment by chiphoi** [comment]
> Since Version 0.6.19 the oldest cover-version of a series appears in the series grid-view 👎

We also need a way to change a cover of a series, instead of sometime picking the cover of a 1st book, sometime picking the last one.

**Comment by gelbphoenix** [comment]
> > Since Version 0.6.19 the oldest cover-version of a series appears in the series grid-view 👎
> 
> We also need a way to change a cover of a series, instead of sometime picking the cover of a 1st book, sometime picking the last one.

Agreed. Especially as some series do have a different or a very known cover for the whole series.

**Comment by gelbphoenix** [comment]
> > Since Version 0.6.19 the oldest cover-version of a series appears in the series grid-view 👎
> 
> We also need a way to change a cover of a series, instead of sometime picking the cover of a 1st book, sometime picking the last one.

Agreed. Especially as some series do have a different or a very known cover for the whole series.
But is that implementation even possible, @OzzieIsaacs?

**Comment by TheDarkSpock** [comment]
+1 for changing cover of a series please. I hate that all of mine show the last book in a series.
